### PR TITLE
Fixed the pagerando specific tests.

### PR DESCRIPTION
**Pull request by Saldivarcher**
This is a fix for the `pagerando` tests that seemed to be failing due to syntax change, and other various reasons.

My main concern was for the `varargs` test, `opt` seemed to be putting the functions in differing places like so:
```
define void @real_va_type(...) #0 {
  %1 = alloca %struct.va_list
  %2 = bitcast %struct.va_list* %1 to i8*
  call void @llvm.va_start(i8* %2)
  call void @"real_va_type$$origva"(%struct.va_list* %1)
  %3 = bitcast %struct.va_list* %1 to i8*
  call void @llvm.va_end(i8* %3)
  ret void
}

; Function Attrs: noinline optsize
define void @copy(...) #0 {
  %1 = alloca i8
  call void @llvm.va_start(i8* %1)
  call void @"copy$$origva"(i8* %1)
  call void @llvm.va_end(i8* %1)
  ret void
}

; Function Attrs: noinline optsize
define void @multiple_starts(...) #0 {
  %1 = alloca i8
  call void @llvm.va_start(i8* %1)
  call void @"multiple_starts$$origva"(i8* %1)
  call void @llvm.va_end(i8* %1)
  ret void
}

; Function Attrs: noinline optsize
define i32 @varags(...) #0 {
  %1 = alloca i8
  call void @llvm.va_start(i8* %1)
  %2 = call i32 @"varags$$origva"(i8* %1)
  call void @llvm.va_end(i8* %1)
  ret i32 %2
}
```
So, is using `CHECK-DAG`, the correct way of handling that?

### CODE REVIEW

**Comment by rinon** [comment]
I haven't gone through closely, will do that later. However, I can tell you right now that CHECK-DAG isn't what you want for the varargs. All CHECK-DAG lines can be re-ordered, until hitting a non-DAG check. So... in this case, using CHECK-DAG on all the lines means they might be seen in any order, which seems to loose of a constraint to me.

File: test/CodeGen/AArch64/pagerando-isel.ll [hunk 1]
```diff
@@ -5,48 +5,48 @@
...
-; CHECK-DAG: %vreg{{[0-9]+}}<def,tied1> = MOVKXi %vreg{{[0-9]+}}<tied0>, <ga:@legacy>[TF=165], 16
-; CHECK-DAG: %vreg{{[0-9]+}}<def,tied1> = MOVKXi %vreg{{[0-9]+}}<tied0>, <ga:@legacy>[TF=164], 32
-; CHECK-DAG: [[LEGACY_GOTOFF:%vreg[0-9]+]]<def,tied1> = MOVKXi %vreg{{[0-9]+}}<tied0>, <ga:@legacy>[TF=35], 48
-; CHECK: [[LEGACY:%vreg[0-9]+]]<def> = LOADgotr [[GOT]], [[LEGACY_GOTOFF]]<kill>
-; CHECK: BLR [[LEGACY]]<kill>
+define void @user() pagerando {
+; CHECK-DAG: [[POT:%[0-9]+]]:gpr64 = COPY $x20
+; CHECK-DAG: [[GOT:%[0-9]+]]:gpr64common = LOADpot [[POT]]:gpr64, 0
+
+; CHECK-DAG: %{{[0-9]+}}:gpr64 = MOVZXi target-flags(aarch64-g0, aarch64-nc, <unknown bitmask target flag>) @legacy, 0
```
Comment (rinon) [inline test/CodeGen/AArch64/pagerando-isel.ll:17]: We should probably fix the unknown bitmask target flags. Should be a separate commit though.

File: test/CodeGen/ARM/pagerando-isel.ll [hunk 2]
```diff
@@ -5,47 +5,51 @@
...
+; CHECK: [[TEST:%[0-9]+]]:gpr = LDRrs killed [[WRAPPER]]:gpr, killed [[WRAPPER_TEST]]
+; CHECK: BLX killed [[TEST]]:gpr
   call void @wrapper()
-
-; CHECK: [[BINNED_POTOFF:%vreg[0-9]+]]<def> = LDRi12 <cp#2>, 0, pred:14, pred:%noreg
-; CHECK: [[BINNED_BIN:%vreg[0-9]+]]<def> = LDRrs [[POT]], [[BINNED_POTOFF]]<kill>, 0, pred:14, pred:%noreg
-; CHECK: [[BINNED_BINOFF:%vreg[0-9]+]]<def> = LDRi12 <cp#3>, 0, pred:14, pred:%noreg
-; CHECK: [[BINNED:%vreg[0-9]+]]<def> = ADDrr [[BINNED_BIN]]<kill>, [[BINNED_BINOFF]]<kill>, pred:14, pred:%noreg, opt:%noreg
-; CHECK: BLX [[BINNED]]<kill>
+
```
Comment (rinon) [inline test/CodeGen/ARM/pagerando-isel.ll]: Trailing whitespace

File: test/CodeGen/AArch64/pagerando-isel.ll [hunk 3]
```diff
@@ -5,48 +5,48 @@
 @internal_var = internal global i32 0
 
 define void @legacy() { ret void }
-define void @wrapper() pagerando_wrapper { ret void }
-define hidden void @binned() pagerando_binned { ret void }
+define void @wrapper() pagerando { ret void }
```
Comment (rinon) [inline test/CodeGen/AArch64/pagerando-isel.ll]: Wrapper functions don't get the pagerando attribute applied (see the PagerandoWrappers tests). This applies to everything called a "wrapper." The backend specific tests are after binning, so we are simulating a wrapper (in .text) and a wrapped, binned function.

File: test/Transforms/PagerandoWrappers/varargs.ll [hunk 4]
```diff
@@ -1,56 +1,68 @@
...
 
-; CHECK-LABEL: define internal void @user()
-; CHECK-NEXT:    call i32 (...) @varags(i32 13, i32 37)
-
 define internal void @user() {
   call i32 (...) @varags(i32 13, i32 37)
   ret void
 }
+; CHECK-LABEL: define void @real_va_type(...)
+; CHECK:    %1 = alloca %struct.va_list
```
Comment (rinon) [inline test/Transforms/PagerandoWrappers/varargs.ll]: Why did you change `CHECK-NEXT` into `CHECK` directives? If there's a reason that's ok, but I just would like to know why.

File: test/CodeGen/ARM/pagerando-optimizer.ll [hunk 5]
```diff
@@ -1,15 +1,17 @@
 ; RUN: llc < %s -mtriple=armv7-linux -relocation-model=pip -o - | FileCheck %s
 
-; CHECK-LABEL: .text
+; CHECK-LABEL: .section .text.bin_1
+; CHECK-LABEL: .type wrapper
 ; CHECK-LABEL: wrapper:
-define void @wrapper() pagerando_wrapper { ret void }
+define void @wrapper() pagerando { ret void }
```
Comment (rinon) [inline test/CodeGen/ARM/pagerando-optimizer.ll]: Again, this shouldn't have the pagerando attribute (see above).

File: test/CodeGen/ARM/pagerando-optimizer.ll [hunk 6]
```diff
@@ -1,15 +1,17 @@
 ; RUN: llc < %s -mtriple=armv7-linux -relocation-model=pip -o - | FileCheck %s
 
-; CHECK-LABEL: .text
+; CHECK-LABEL: .section .text.bin_1
+; CHECK-LABEL: .type wrapper
```
Comment (rinon) [inline test/CodeGen/ARM/pagerando-optimizer.ll]: I don't think we need these `.type` lines. I think they're for type-based CFI?

File: test/CodeGen/AArch64/pagerando-optimizer.ll [hunk 7]
```diff
@@ -1,19 +1,18 @@
 ; RUN: llc < %s -march=aarch64 -relocation-model=pip -o - | FileCheck %s
 
-; CHECK-LABEL: .text
+; CHECK-LABEL: section .text.bin_1
 ; CHECK-LABEL: wrapper:
-define void @wrapper() pagerando_wrapper { ret void }
+define void @wrapper() pagerando { ret void }
```
Comment (rinon) [inline test/CodeGen/AArch64/pagerando-optimizer.ll]: No pagerando attr (see above)

**Review by rinon — COMMENTED** [review]
Almost good to go, just a few things to clean up.

File: test/Transforms/PagerandoWrappers/varargs.ll (hunk 4 above)
Comment (Saldivarcher) [reply test/Transforms/PagerandoWrappers/varargs.ll]: Oops, you're right it should be `CHECK-NEXT`.

File: test/CodeGen/AArch64/pagerando-isel.ll (hunk 3 above)
Comment (Saldivarcher) [reply test/CodeGen/AArch64/pagerando-isel.ll]: Okay, I get it now. Thank you for the explanation :smile:

File: test/Transforms/PagerandoWrappers/varargs.ll [hunk 8]
```diff
@@ -1,18 +1,39 @@
 ; RUN: opt < %s -pagerando-wrappers -S | FileCheck %s
 
+; CHECK-LABEL: %struct.va_list = type { i8*, i8*, i8*, i32, i32 }
```
Comment (rinon) [inline test/Transforms/PagerandoWrappers/varargs.ll]: No need to add this. We want the test to be as generic and minimal as possible while still checking for the functionality we're testing.

File: test/CodeGen/ARM/pagerando-isel.ll (hunk 2 above)
Comment (rinon) [reply test/CodeGen/ARM/pagerando-isel.ll]: Remove this whitespace and the changes LGTM.

**Review by rinon — APPROVED** [review]
